ank/src/cli_commands/server_connection.rs

```diff
@@ -0,0 +1,800 @@
+// Copyright (c) 2023 Elektrobit Automotive GmbH
+//
+// This program and the accompanying materials are made available under the
+// terms of the Apache License, Version 2.0 which is available at
+// https://www.apache.org/licenses/LICENSE-2.0.
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations
+// under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{mem::take, time::Duration};
+
+use crate::{output_and_error, output_debug};
+use common::communications_client::CommunicationsClient;
+use common::to_server_interface::ToServer;
+use common::{
+    commands::{
+        CompleteStateRequest, Response, ResponseContent, UpdateStateSuccess, UpdateWorkloadState,
+    },
+    from_server_interface::{FromServer, FromServerReceiver},
+    objects::CompleteState,
+    to_server_interface::{ToServerInterface, ToServerSender},
+};
+use grpc::client::GRPCCommunicationsClient;
+#[cfg(test)]
+use mockall::automock;
+use url::Url;
+
+const BUFFER_SIZE: usize = 20;
+const WAIT_TIME_MS: Duration = Duration::from_millis(3000);
+
+pub struct ServerConnection {
+    to_server: ToServerSender,
+    from_server: FromServerReceiver,
+    task: tokio::task::JoinHandle<()>,
+    missed_from_server_messages: Vec<FromServer>,
```

Comment:
The CLI is not going to run in production, but I wonder if we should still protect ourselves against excessive memory usage here by using a ring buffer.
In theory this can only happen if the server does not respond in time and we have a lot of workloads in the system that are constantly changing states.
Using a ring buffer would of course break the functionality. Monitoring the size of the vector and exiting above a certain value would also be a possibility.
Admittedly, this is all very theoretical and will probably never occur. Maybe for now we can just keep it in mind.

Reply:
What about this: we can implement a check somewhere in a code whether the vector has exceeded a threshold or not and print a warning message. If we see this warning message in a future, we will implement clean solution.

Reply:
I think we should keep it as it is. As it is interactive tool, the user will probably stop it if it takes to long. Also it will be run a normal computer and not embedded devices, hence the memory usage is not a huge concern.

If we do something, I would suggest introducing a timeout. This will not solve the problem directly, as many messages in a short time could still lead to the buffer getting big, but the chances are smaller.

Reply:
Maybe you are right, we can keep it as it is. This is also one of the options from Kaloyan. The last suggestion: what about writing here a comment with warning that we have here potential issue? But it is not really necessary.

Reply:
I agree, this is a decision we took and the correct place to document it is the design decisions section of the swdd.